### PR TITLE
disttask: add hook for executor

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #48795 

Problem Summary:

Currently, so many failpoints are added to taskExecutor which make the code in a mess.
We need hook to make the test easier.
Furthermore, hook can be used to do more customized logic.

Why not mock？
1. the Onxxx interface is optional, we just need to implement some of them.

### What changed and how does it work?
Add hook for executor.
```
// Callback is used for task executor.
type Callback interface {
	OnInitBefore(task *proto.Task) error
	OnSubtaskRunBefore(subtask *proto.Subtask) bool
	OnSubtaskRunAfter(subtask *proto.Subtask) error
	OnSubtaskFinishedBefore(subtask *proto.Subtask) error
	OnSubtaskFinishedAfter(subtask *proto.Subtask)
}
```
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/disttask/framework/framework_ha_test.go [hunk 1]
```diff
@@ -16,24 +16,91 @@ package framework_test
...
+	"github.com/pingcap/tidb/pkg/disttask/framework/hook"
+	"github.com/pingcap/tidb/pkg/disttask/framework/proto"
+	"github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor"
 	"github.com/pingcap/tidb/pkg/disttask/framework/testutil"
+	"github.com/pingcap/tidb/pkg/domain/infosync"
 	"github.com/stretchr/testify/require"
 )
 
+func registerHAHook() {
+	// init hk.
```
Comment (ywqzzy) [inline pkg/disttask/framework/framework_ha_test.go:31]:
```suggestion
```

File: pkg/disttask/framework/taskexecutor/task_executor.go [hunk 2]
```diff
@@ -636,3 +571,40 @@ func (s *BaseTaskExecutor) updateErrorToSubtask(ctx context.Context, taskID int6
...
+
+func (s *BaseTaskExecutor) onSubtaskFinishedBefore(subtask *proto.Subtask) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.mu.hook.OnSubtaskFinishedBefore(subtask)
+}
+
+func (s *BaseTaskExecutor) onSubtaskFinishedAfter(subtask *proto.Subtask) {
+	s.mu.Lock()
```
Comment (ywqzzy) [inline pkg/disttask/framework/taskexecutor/task_executor.go:606]: Maybe the mutex is useless, I will fire another pr to remove the mutex

Comment (ywqzzy) [reply pkg/disttask/framework/taskexecutor/task_executor.go:606]: > The mutex in taskExecutor is useless, I will fire another pr to remove the mutex

**Comment by okJiang** [comment]
I am so sorry, I don't the code is more readable after this pr in fact. And the scalability is not very well.

For example, the interface

```go
// Callback is used for task executor.
type Callback interface {
	// OnInitBefore called before taskExecutor.Init.
	OnInitBefore(task *proto.Task) error
	// OnSubtaskRunBefore called before task executor run one subtask.
	OnSubtaskRunBefore(subtask *proto.Subtask) bool
	// OnSubtaskRunAfter called after task executor run one subtask.
	OnSubtaskRunAfter(subtask *proto.Subtask) error
	// OnSubtaskFinishedBefore called before mark one subtask as finished.
	OnSubtaskFinishedBefore(subtask *proto.Subtask) error
	// OnSubtaskFinishedAfter called after mark one subtask as finished.
	OnSubtaskFinishedAfter(subtask *proto.Subtask)
}
```

The value returned is highly customized, such as `error`, `bool`. The next person that need add some code to the hook need figure out what is it firstly.

**Comment by ywqzzy** [comment]
> I am so sorry, I don't the code is more readable after this pr in fact. And the scalability is not very well.
> 
> For example, the interface
> 
> ```go
> // Callback is used for task executor.
> type Callback interface {
> 	// OnInitBefore called before taskExecutor.Init.
> 	OnInitBefore(task *proto.Task) error
> 	// OnSubtaskRunBefore called before task executor run one subtask.
> 	OnSubtaskRunBefore(subtask *proto.Subtask) bool
> 	// OnSubtaskRunAfter called after task executor run one subtask.
> 	OnSubtaskRunAfter(subtask *proto.Subtask) error
> 	// OnSubtaskFinishedBefore called before mark one subtask as finished.
> 	OnSubtaskFinishedBefore(subtask *proto.Subtask) error
> 	// OnSubtaskFinishedAfter called after mark one subtask as finished.
> 	OnSubtaskFinishedAfter(subtask *proto.Subtask)
> }
> ```
> 
> The value returned is highly customized, such as `error`, `bool`. The next person that need add some code to the hook need figure out what is it firstly. It is not friendly.

See pkg/ddl/callback.go
The interface can be changed.
Too many failpoint will be make the code horrible.

**Comment by ywqzzy** [comment]
> > I am so sorry, I don't the code is more readable after this pr in fact. And the scalability is not very well.
> > For example, the interface
> > ```go
> > // Callback is used for task executor.
> > type Callback interface {
> > 	// OnInitBefore called before taskExecutor.Init.
> > 	OnInitBefore(task *proto.Task) error
> > 	// OnSubtaskRunBefore called before task executor run one subtask.
> > 	OnSubtaskRunBefore(subtask *proto.Subtask) bool
> > 	// OnSubtaskRunAfter called after task executor run one subtask.
> > 	OnSubtaskRunAfter(subtask *proto.Subtask) error
> > 	// OnSubtaskFinishedBefore called before mark one subtask as finished.
> > 	OnSubtaskFinishedBefore(subtask *proto.Subtask) error
> > 	// OnSubtaskFinishedAfter called after mark one subtask as finished.
> > 	OnSubtaskFinishedAfter(subtask *proto.Subtask)
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The value returned is highly customized, such as `error`, `bool`. The next person that need add some code to the hook need figure out what is it firstly. It is not friendly.
> 
> See pkg/ddl/callback.go The interface can be changed. Too many failpoint will be make the code horrible.

And IMHO, if the next person don't know the interface well, he won't have the ability to change the code since he don't know the error handling process of disttask framework and don't know the task executor running process.

**Comment by ywqzzy** [comment]
> > > I am so sorry, I don't the code is more readable after this pr in fact. And the scalability is not very well.
> > > For example, the interface
> > > ```go
> > > // Callback is used for task executor.
> > > type Callback interface {
> > > 	// OnInitBefore called before taskExecutor.Init.
> > > 	OnInitBefore(task *proto.Task) error
> > > 	// OnSubtaskRunBefore called before task executor run one subtask.
> > > 	OnSubtaskRunBefore(subtask *proto.Subtask) bool
> > > 	// OnSubtaskRunAfter called after task executor run one subtask.
> > > 	OnSubtaskRunAfter(subtask *proto.Subtask) error
> > > 	// OnSubtaskFinishedBefore called before mark one subtask as finished.
> > > 	OnSubtaskFinishedBefore(subtask *proto.Subtask) error
> > > 	// OnSubtaskFinishedAfter called after mark one subtask as finished.
> > > 	OnSubtaskFinishedAfter(subtask *proto.Subtask)
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The value returned is highly customized, such as `error`, `bool`. The next person that need add some code to the hook need figure out what is it firstly. It is not friendly.
> > 
> > 
> > See pkg/ddl/callback.go The interface can be changed. Too many failpoint will be make the code horrible.
> 
> And IMHO, if the next person don't know the interface well, he won't have the ability to change the code since he don't know the error handling process of disttask framework and don't know the task executor running process.

Only if we deep dive into the error handling process can we know where to inject a hook of a failpoint.

**Comment by ywqzzy** [comment]
> > > > I am so sorry, I don't the code is more readable after this pr in fact. And the scalability is not very well.
> > > > For example, the interface
> > > > ```go
> > > > // Callback is used for task executor.
> > > > type Callback interface {
> > > > 	// OnInitBefore called before taskExecutor.Init.
> > > > 	OnInitBefore(task *proto.Task) error
> > > > 	// OnSubtaskRunBefore called before task executor run one subtask.
> > > > 	OnSubtaskRunBefore(subtask *proto.Subtask) bool
> > > > 	// OnSubtaskRunAfter called after task executor run one subtask.
> > > > 	OnSubtaskRunAfter(subtask *proto.Subtask) error
> > > > 	// OnSubtaskFinishedBefore called before mark one subtask as finished.
> > > > 	OnSubtaskFinishedBefore(subtask *proto.Subtask) error
> > > > 	// OnSubtaskFinishedAfter called after mark one subtask as finished.
> > > > 	OnSubtaskFinishedAfter(subtask *proto.Subtask)
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > The value returned is highly customized, such as `error`, `bool`. The next person that need add some code to the hook need figure out what is it firstly. It is not friendly.
> > > 
> > > 
> > > See pkg/ddl/callback.go The interface can be changed. Too many failpoint will be make the code horrible.
> > 
> > 
> > And IMHO, if the next person don't know the interface well, he won't have the ability to change the code since he don't know the error handling process of disttask framework and don't know the task executor running process.
> 
> Only if we deep dive into the error handling process can we know where to inject a hook or a failpoint.

**Comment by ywqzzy** [comment]
> > > > > I am so sorry, I don't the code is more readable after this pr in fact. And the scalability is not very well.
> > > > > For example, the interface
> > > > > ```go
> > > > > // Callback is used for task executor.
> > > > > type Callback interface {
> > > > > 	// OnInitBefore called before taskExecutor.Init.
> > > > > 	OnInitBefore(task *proto.Task) error
> > > > > 	// OnSubtaskRunBefore called before task executor run one subtask.
> > > > > 	OnSubtaskRunBefore(subtask *proto.Subtask) bool
> > > > > 	// OnSubtaskRunAfter called after task executor run one subtask.
> > > > > 	OnSubtaskRunAfter(subtask *proto.Subtask) error
> > > > > 	// OnSubtaskFinishedBefore called before mark one subtask as finished.
> > > > > 	OnSubtaskFinishedBefore(subtask *proto.Subtask) error
> > > > > 	// OnSubtaskFinishedAfter called after mark one subtask as finished.
> > > > > 	OnSubtaskFinishedAfter(subtask *proto.Subtask)
> > > > > }
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > The value returned is highly customized, such as `error`, `bool`. The next person that need add some code to the hook need figure out what is it firstly. It is not friendly.
> > > > 
> > > > 
> > > > See pkg/ddl/callback.go The interface can be changed. Too many failpoint will be make the code horrible.
> > > 
> > > 
> > > And IMHO, if the next person don't know the interface well, he won't have the ability to change the code since he don't know the error handling process of disttask framework and don't know the task executor running process.
> > 
> > 
> > Only if we deep dive into the error handling process can we know where to inject a hook or a failpoint.

And all the hook interface won't be changed frequently. We recommend users to add interface instead of change them.

File: pkg/disttask/framework/hook/callback.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
+// Copyright 2015 PingCAP, Inc.
```
Comment (ywqzzy) [inline pkg/disttask/framework/hook/callback.go:1]:
```suggestion
// Copyright 2023 PingCAP, Inc.
```

**Comment by D3Hunter** [comment]
hasn't reviewed the code, some my own opinions:

> Currently, so many failpoints are added to taskExecutor which make the code in a mess.
> We need hook to make the test easier.

maybe we can try to refactor the code, such as using interface or function pointer to mock, it's much easier to mock than hooks

some times we want to control code flow using failpoint, like return specific data or error, or skip some part of the code, callback doesn't help in this

> Furthermore, hook can be used to do more customized logic.

disttask is not a event-based system, i don't think we should provide those facilities too early before they're actually needed.

**Comment by ywqzzy** [comment]
> maybe we can try to refactor the code, such as using interface or function pointer to mock, it's much easier to mock than hooks
1. The hook is just like the function ptr.
2. some of the failpoints are kept here.

**Comment by ywqzzy** [comment]
> interface or function pointer to mock,

**Comment by D3Hunter** [comment]
> The hook is just like the function ptr, we just init it if needed.

hook is too intrusive. for function ptr:

https://github.com/pingcap/tidb/blob/f50e89450e388313574e3b955c696b4903569dac/pkg/disttask/importinto/subtask_executor.go#L62

**Pull request closed without merging**